api/src/main/java/me/lokka30/treasury/api/economy/currency/Currency.java

```diff
@@ -12,69 +12,224 @@
 
 package me.lokka30.treasury.api.economy.currency;
 
-import org.jetbrains.annotations.NotNull;
-import org.jetbrains.annotations.Nullable;
-
 import java.util.Locale;
+import java.util.Objects;
 import java.util.UUID;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
 
 /**
- * @author lokka30, Geolykt
- * @since v1.0.0
  * One of Treasury's core features is multi-currency support.
  * This allows economy providers and plugins to use different
  * currencies for different transactions, e.g. a daily reward
  * plugin can award players 'Tokens', but a job plugin
  * can award players 'Dollars'. Facilitates great customisability.
+ *
+ * @author lokka30, Geolykt, MrIvanPlays
+ * @since v1.0.0
  */
-@SuppressWarnings("unused")
-public interface Currency {
+public final class Currency {
```

Comment:
I don't think this class should be final as I explained here. 
https://github.com/lokka30/Treasury/issues/49#issuecomment-986275101
With the example of simulating money, lets say my plugin simulates cryptocurrencyes:
I have bitcoin and etherium, depening on how much is buyed/selled i want these currencyes to have a higer/lower value, i can modify these by changing its conversionCoefficient, I could do this in several ways:
1- Override currency so I can change the returned value as I want
2- Change the value of conversionCoefficient when necesary (its final so it's kinda hard)
3- conversionCoefficient is a supplier that I provide (?)
Right now I couldn't simulate this because I can't modify the conversionCoefficient